applications/doodle/src/doodle_inbound_listener.erl

```diff
@@ -0,0 +1,187 @@
+%%% @copyright (C) 2013, 2600Hz
+%%% @doc
+%%%
+%%% @end
+%%% @contributors
+%%%-------------------------------------------------------------------
+-module(doodle_inbound_listener).
+
+-behaviour(gen_listener).
+
+-export([start_link/0]).
+-export([init/1
+         ,handle_call/3
+         ,handle_cast/2
+         ,handle_info/2
+         ,handle_event/2
+         ,terminate/2
+         ,code_change/3
+        ]).
+
+-include("doodle.hrl").
+
+-record(state, {}).
+
+-define(DEFAULT_EXCHANGE, <<"sms">>).
+-define(DEFAULT_EXCHANGE_TYPE, <<"topic">>).
+-define(DEFAULT_EXCHANGE_OPTIONS, []).
+
+-define(DOODLE_INBOUND_EXCHANGE, whapps_config:get_ne_binary(?CONFIG_CAT, <<"inbound_exchange">>, ?DEFAULT_EXCHANGE)).
+-define(DOODLE_INBOUND_EXCHANGE_TYPE, whapps_config:get_ne_binary(?CONFIG_CAT, <<"inbound_exchange_type">>, ?DEFAULT_EXCHANGE_TYPE)).
+-define(DOODLE_INBOUND_EXCHANGE_OPTIONS,  whapps_config:get(?CONFIG_CAT, <<"inbound_exchange_options">>, ?DEFAULT_EXCHANGE_OPTIONS)).
+        
+
+-define(BINDINGS(Ex), [{'sms', [{'exchange', Ex}
+                                ,{'restrict_to', ['inbound']}]}
+                       ,{'self', []}
+                      ]).
+-define(RESPONDERS, [{'doodle_inbound_handler',[{<<"message">>, <<"inbound">>}]}]).
+
+-define(QUEUE_NAME, <<"smsc_inbound_queue">>).
+-define(QUEUE_OPTIONS, [{'exclusive', 'false'}
+                        ,{'durable', 'true'}
+                        ,{'auto_delete', 'false'}
+                        ,{'arguments', [{<<"x-message-ttl">>, 'infinity'}
+                                        ,{<<"x-max-length">>, 'infinity'}
+                                       ]}
+                       ]).
+-define(CONSUME_OPTIONS, [{'exclusive', 'false'}
+                          ,{'no_ack', 'false'}
+                         ]).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the server
+%%
+%% @spec start_link() -> {ok, Pid} | ignore | {error, Error}
+%% @end
+%%--------------------------------------------------------------------
+start_link() ->
+    Exchange = ?DOODLE_INBOUND_EXCHANGE,
+    Type = ?DOODLE_INBOUND_EXCHANGE_TYPE,
+    Options = ?DOODLE_INBOUND_EXCHANGE_OPTIONS,
+    ED = amqp_util:declare_exchange(Exchange, Type, Options),
+    gen_listener:start_link({'local', ?MODULE}
+                            ,?MODULE
+                            ,[{'bindings', ?BINDINGS(Exchange)}
+                              ,{'responders', ?RESPONDERS}
+                              ,{'queue_name', ?QUEUE_NAME}       % optional to include
+                              ,{'queue_options', ?QUEUE_OPTIONS} % optional to include
+                              ,{'consume_options', ?CONSUME_OPTIONS} % optional to include
+                              ,{'declare_exchanges', [ED]}
```

Comment:
In fact, I think this should probably be in the wapi_* modules only. So wapi_sms should take the 2- or 3-tuple list and declare the exchanges itself, and leave gen_listener out of it.